frontend/packages/sdk/src/contracts/PinkBlueprint.ts

```diff
@@ -271,37 +309,110 @@ export class PinkBlueprintPromise {
+        return phalaTypes.createType<InkQueryError>('InkQueryError', response.result.asErr.toHex())
       }
-      const result = api.createType<ContractInstantiateResult>(
+      const result = phalaTypes.createType<ContractInstantiateResult>(
         'ContractInstantiateResult',
         response.result.asOk.asInkMessageReturn.toHex()
       )
       ;(result as PinkContractInstantiateResult).salt = salt
+
+      if (result.result.isErr) {
+        const err = result.result.asErr
+        if (err.isModule && err.asModule.index.toNumber() === 4) {
+          const contractError = phalaTypes.createType('ContractError', result.result.asErr.asModule.error)
+          throw new Error(`Estimation failed: ${contractError.toHuman()}`)
+        }
+        throw new Error('Estimation failed: ' + JSON.stringify(result.result.asErr.toHuman()))
+      }
+
       return result
     }
 
     return {
       send: this._decorateMethod((origin: string | AccountId | Uint8Array) => from(inkQueryInternal(origin))),
     }
   }
+
+  async #send(constructorOrId: string, options: PinkBlueprintSendOptions, ...args: unknown[]) {
+    const { cert: userCert, ...rest } = options
+    const txOptions: PinkBlueprintOptions = {
+      gasLimit: options.gasLimit,
+      value: options.value,
+      storageDepositLimit: options.storageDepositLimit,
+    }
+
+    const tx = this.#tx[constructorOrId]
+    if (!tx) {
+      throw new Error(`Constructor not found: ${constructorOrId}`)
+    }
+
+    const address = 'signer' in rest ? rest.address : rest.pair.address
+    const cert = userCert || (await this.phatRegistry.getAnonymousCert())
+    const estimate = this.#query[constructorOrId]
+    if (!estimate) {
+      throw new Error(`Constructor not found: ${constructorOrId}`)
+    }
+
+    const gasPrice = this.phatRegistry.clusterInfo?.gasPrice
+    const depositPerByte = this.phatRegistry.clusterInfo?.depositPerByte
+    if (!gasPrice || !depositPerByte) {
+      throw new Error('No Gas Price or deposit Per Byte from cluster info.')
+    }
+
+    //
+    // Even the estimate result already large then the real cost, we still need deposit more
+    // for the peak usage.
+    //
+    const mutlipiler = 1.05
+
+    const msg = this.abi.findConstructor(constructorOrId).toU8a(args)
+    const deposit = depositPerByte.mul(new BN(msg.length * mutlipiler))
+
+    const [clusterBalance, onchainBalance, { gasRequired, storageDeposit }] = await Promise.all([
+      this.phatRegistry.getClusterBalance(address),
+      this.api.query.system.account<FrameSystemAccountInfo>(address),
+      // We estimating the gas & storage deposit cost with deposit propose.
+      estimate(cert.address, { cert, deposit }, ...args),
+    ])
+
+    // We calculate the minimal cost based on the estimated result. Even the estimated result already higher than
+    // the real-cost, we still need to deposit more for the peak usage.
```

Comment:
Because you forget the `gasPrice`.
It should be `minRequired = gasRequired * gasPrice + storageDeposit`.